src/plugin.js

```diff
@@ -1,96 +1,156 @@
-const path = require('path');
 const lpModule = import('load-plugin');
 ('use strict');
 
 /**
- * Finds, registers and loads plugins used by git-proxy
+ * Checks if the given object or any of its prototypes has the 'isGitProxyPlugin' property set to true.
+ * @param {Object} obj - The object to check.
+ * @param {string} propertyName - The property name to check for. Default is 'isGitProxyPlugin'.
+ * @return {boolean} - True if the object or any of its prototypes has the 'isGitProxyPlugin' property set to true, false otherwise.
+ */
+function isCompatiblePlugin(obj, propertyName = 'isGitProxyPlugin') {
+  while (obj != null) {
+    if (Object.prototype.hasOwnProperty.call(obj, propertyName) &&
+      obj.isGitProxyPlugin &&
+      Object.keys(obj).includes('exec')) {
+      return true;
+    }
+    obj = Object.getPrototypeOf(obj);
+  }
+  return false;
+}
+
+/**
+ * @typedef PluginTypeResult
+ * @property {ProxyPlugin[]} pushPlugins - List of push plugins
+ * @property {ProxyPlugin[]} pullPlugins - List of pull plugins
+ */
+
+/**
+ * Registers and loads plugins used by git-proxy
  */
 class PluginLoader {
+  /**
+   * @property {Promise} load - A Promise that begins loading plugins from a list of modules. Callers must run `await loader.load` to load plugins.
+   */
+  load;
+  /**
+   * This property is not used in production code. It is exposed for testing purposes.
+   * @property {Promise} ready - A Promise that resolves when all plugins have been loaded.
+   */
+  ready;
   /**
    * Initialize PluginLoader with candidates modules (node_modules or relative
    * file paths).
-   * @param {Array.<string>} names List of Node module/package names to load.
-   * @param {Array.<string>} paths List of file paths to load modules from.
+   * @param {Array.<string>} targets List of Node module package names or files to load.
    */
-  constructor(names, paths) {
-    this.names = names;
-    this.paths = paths;
+  constructor(targets) {
+    this.targets = targets;
     /**
-     * @type {Array.<ProxyPlugin>} List of ProxyPlugin objects loaded.
+     * @type {ProxyPlugin[]} List of loaded ProxyPlugins
      * @public
      */
-    this.plugins = [];
+    this.pushPlugins = [];
+    this.pullPlugins = [];
+    if (this.targets.length === 0) {
+      console.log('No plugins configured'); // TODO: log.debug()
```

Comment:
Why `TODO`?

Reply:
Reference for #399 - replace with a proper logger.